### PR TITLE
context: detect if README is required

### PR DESCRIPTION
**Pull request by stefanhengl**
This adds a thin layer to expand a query with keywords which are not explicitly mentioned. As a first application, we detect if a README should be returned. See /vscode/src/chat/chat-view/context.ts for how a similar logic is implemented in the client.

Test plan:
- new and updated unit tests
- Cody web client: I tested the "what does this codebase do?" for `sourcegraph/sourcegraph` and `sourcegraph/zoekt` as selected context. For Zoekt we return `/README.md` in the top 3, which leads to a good answer. The Sourcegraph repo has many README files and our ranking ranks other README files higher, for example if they contain the string "README", which leads to a vague answer.

Addresses SPLF-94

### CODE REVIEW

File: internal/search/codycontext/query_parser.go [hunk 1]
```diff
@@ -194,3 +197,68 @@ func findSymbols(patterns []string) []string {
...
+}
+
+var questionSignifiers = []string{
+	"what",
+	"how",
+	"describe",
+	"explain",
+}
+
+func needsReadmeContext(input string) bool {
```
Comment (jtibshirani) [inline internal/search/codycontext/query_parser.go:228]: It'd be really nice to also match queries containing only the repo name, like
* "what does batch-change-buddy/jtest do?" 
* "what does go-ctags do?"

Do you think it'd be possible to pull out the repo names and also match against these? They would count as "project signifiers".

**Review by jtibshirani — COMMENTED** [review]
> The Sourcegraph repo has many README files and our ranking ranks other README files higher, for example if they contain the string "README", which leads to a vague answer.

It feels like we should consistently match the top-level README. Do you plan to follow-up with a fix? Maybe boosting exact matches on filename higher in Zoekt? Brainstorming ...

```diff
--- a/score.go
+++ b/score.go
@@ -118,8 +118,13 @@ func calculateTermFrequency(cands []*candidateMatch, df termDocumentFrequency) m
        termFreqs := map[string]int{}
        for _, cand := range cands {
                term := string(cand.substrLowered)
+               length := uint32(len(cand.substrLowered))
                if cand.fileName {
-                       termFreqs[term] += 5
+                       if cand.byteOffset == 0 && cand.byteMatchSz == length {
+                               termFreqs[term] += 5
+                       } else {
+                               termFreqs[term] += 2
+                       }
                } else {
                        termFreqs[term]++
                }
```

File: internal/search/codycontext/query_parser.go [hunk 2]
```diff
@@ -33,22 +33,25 @@ func parseQuery(queryString string) (*contextQuery, error) {
 	symbols := findSymbols(patterns)
 	transformedPatterns := transformPatterns(patterns)
 
-	// To maintain decent latency, limit the number of patterns we search.
-	if len(transformedPatterns) > maxTransformedPatterns {
-		transformedPatterns = transformedPatterns[:maxTransformedPatterns]
+	entities := extractEntities(queryString)
+	keywords := append(transformedPatterns, entities...)
```
[the remark this answers is not in the document]
Comment (stefanhengl) [reply internal/search/codycontext/query_parser.go:37]: Interesting, I thought of `transformedPatterns` as having higher confidence. We can switch it around and observe how it behaves IRL.

**Comment by stefanhengl** [comment]
> It feels like we should consistently match the top-level README. Do you plan to follow-up with a fix?

Yes, definitely. Your suggestions looks promising. I will take a look. I wonder whether we could add something that prioritizes shorter paths in general, not just exact matches.

File: internal/search/codycontext/query_parser.go (hunk 1 above)
Comment (stefanhengl) [reply internal/search/codycontext/query_parser.go:225]: I added a regex to match project names on a best-effort basis. I added a couple of tests to check corner cases.

Comment (jtibshirani) [reply internal/search/codycontext/query_parser.go:225]: Couldn't we pull out the repo names being searched and match against those? They are available in the query string.

Comment (stefanhengl) [reply internal/search/codycontext/query_parser.go:225]: I guess I don't understand exactly what you mean with "Couldn't we pull out the repo names being searched and match against those?". I don't think we can reliably identify "go-ctags" as repository unless the user has explicitly selected it as context. We could check the DB but that is quite expensive.

I will remove the repo logic for now and follow-up once we chatted a bit more about this.

**Comment by jtibshirani** [comment]
> I guess I don't understand exactly what you mean with "Couldn't we pull out the repo names being searched and match against those?"

Cody context searches are always guaranteed to include `repo` filters, so I was thinking we could check for those and pull out the repo names. But I am +1 that we removed it, I like the "rewrite the initial query" option better (https://linear.app/sourcegraph/issue/SPLF-98/rewrite-mentioned-entities-to-natural-language).
